### PR TITLE
chore(eslint): enables recommended eslint rules for `testing-library` and `jest-dom`

### LINKED ISSUE #42229: Use react testing library plugins and associated eslint plugins in the client
**Is your feature request related to a problem? Please describe.**
As mentioned by @huyenltnguyen [here](https://github.com/freeCodeCamp/freeCodeCamp/pull/42221/files#r637459504) there are some useful eslint plugins (`eslint-plugin-testing-library` and `eslint-plugin-jest-dom`) and testing library plugins `@testing-library/user-event` and `@testing-library/dom` that could help improve our client tests.

**Describe the solution you'd like**
To use these plugins for the client and fix any issue they highlight.

### PR DESCRIPTION
**Pull request by anirudh-modi**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #42229 

The PR enables the default recommended rules for both `eslint-plugin-testing-library` and `eslint-plugin-jest-dom` which one can extend by this

- https://github.com/testing-library/eslint-plugin-jest-dom#recommended-configuration
- https://github.com/testing-library/eslint-plugin-testing-library#react

### CODE REVIEW

**Comment by anirudh-modi** [comment]
## Impact areas
Below are the list of recommended rules which are violated in the mentioned files.

### client/src/components/Header/Header.test.js
- `testing-library/render-result-naming-convention` 

### client/src/components/Intro/Intro.test.js
- `testing-library/await-async-query` 

### client/src/components/app-mount-notifier.test.tsx
- `testing-library/no-node-access`

### client/src/components/formHelpers/Form.test.js
- `testing-library/prefer-screen-queries`

### client/src/components/formHelpers/block-save-button.test.tsx
- `testing-library/prefer-screen-queries`

### client/src/components/helpers/loader.test.tsx
- `testing-library/no-node-access`

### client/src/components/landing/Landing.test.js
- `testing-library/render-result-naming-convention`

### client/src/components/profile/Profile.test.tsx
- `testing-library/prefer-screen-queries`

### client/src/components/profile/components/HeatMap.test.tsx
- `jest-dom/prefer-to-have-text-content`
- `testing-library/prefer-screen-queries`

### client/src/components/profile/components/TimeLine.test.tsx
- `testing-library/no-container`
- `testing-library/no-node-access`
- `testing-library/prefer-screen-queries`

### client/src/components/search/searchBar/SearchBar.test.tsx
- `testing-library/render-result-naming-convention`

### client/src/components/settings/Certification.test.js
- `testing-library/no-container`
- `testing-library/no-node-access`
- `testing-library/prefer-screen-queries`

### client/src/components/settings/Honesty.test.js
- `testing-library/render-result-naming-convention`
- `testing-library/await-async-query`

### client/src/templates/Challenges/components/completion-modal-body.test.tsx
- `testing-library/no-container`
- `testing-library/no-node-access`
- `jest-dom/prefer-to-have-style`

### client/utils/gatsby/layoutSelector.test.js
- `testing-library/render-result-naming-convention`

**Review by huyenltnguyen — COMMENTED** [review]
@anirudh-modi Thanks for initiating this! 

Let's go through the errors together 💪 

---

https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/app-mount-notifier.test.tsx#L45

I think the proper fix should be converting the query to `getByRole('document')`. However, I found that the library that powers React Testing Library's role mapping, `aria-query`, does not include `html` in the list of elements with the role `document` (see [here](https://github.com/A11yance/aria-query/blob/main/src/etc/roles/literal/documentRole.js)). I don't think there is another way to write this test without `querySelector`. Let's disable the rule in this test for now.

```js
await waitFor(() => {
  /* eslint-disable-next-line testing-library/no-node-access */
  expect(...);
}
```

---
https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/helpers/loader.test.tsx#L16

Firstly, the element the tests are querying is a `div`, so `ByRole` cannot be used here. Secondly, there is a test that checks whether or not a specific class name is applied to the element. This is testing implementation detail, and I think we should try to avoid. Checking whether or not a class name is applied doesn't guarantee that the component will be displayed correctly, as the test would still pass even when we change the CSS of the class. On the other hand, the test would fail if we were to just simply rename the class and without touching the CSS. 

So to resolve errors in these, I agree that we should resort to `testId`, as you suggested earlier. It's not the best option, but better than pulling out the node of the DOM tree directly. Furthermore, we are going to build a new component library and this one will be gone then, so I wouldn't put much effort into this.

---
    
https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/profile/components/TimeLine.test.tsx#L51

There are links and buttons in this component. For querying buttons, let's use

```js
getByRole('button', { name: 'foo' })
``` 

And for links, let's use 

```js
getByRole('link', { name: 'bar' })
```

What these do is they don't just look for a button or a link, but with the given text as well. In a normal setup, if the button has the text "Submit", then "Submit" would be what we look for instead of "foo"/"bar". 

However, in our case, you will need to find them in the component code (TimeLine.tsx). They will be in the following format:

```js
{t('foo')}
```

This `{t()}` is the internationalization setup that you don't have to worry about, just take the string inside it and use it in the tests.

I didn't manage to go through all the errors, but for the first one on line 51, we can resolve it by changing the assertion to:

```js
expect(screen.getByRole('link', { name: 'buttons.view' })).toHaveAttribute(
  'href',
  'https://github.com/freeCodeCamp/freeCodeCamp'
);
```

**Comment by anirudh-modi** [comment]
Hi @huyenltnguyen Thanks a lot for such an intensive research... Really appreciate the help 🙂
These will really help me, let me resolve them individually..

File: client/src/components/settings/Certification.test.js
Comment (anirudh-modi) [inline client/src/components/settings/Certification.test.js:28]: @huyenltnguyen  I found this test and the one above it similar so I had removed the test.. Correct me if I am wrong..

File: client/src/templates/Challenges/components/completion-modal-body.tsx
Comment (anirudh-modi) [inline client/src/templates/Challenges/components/completion-modal-body.tsx:80]: @huyenltnguyen As these were `svg` and `div` I had to use `testId`... I tried using `role=graphics-document` for `svg` as mentioned [here](https://www.w3.org/TR/html-aria/#docconformance)

However, it did not helped..

and where as for the `div` I could not use `text` as the text is on the inner element and we are validating the style on parent element, which forced me to use `testId`

File: client/src/components/settings/Certification.test.js
Comment (anirudh-modi) [inline client/src/components/settings/Certification.test.js:57]: I am bit skeptical of this, ie, if is there a better/more correct way to test the same given the new constraint we have.. Suggestions  are welcome.. 😄

File: client/src/templates/Challenges/components/completion-modal-body.tsx
Comment (huyenltnguyen) [reply client/src/templates/Challenges/components/completion-modal-body.tsx:80]: Yeah the role `graphics-document` doesn't seem to be available in [aria-query](https://github.com/A11yance/aria-query/tree/main/src/etc/roles/literal).

The GreenPass component has a `<title>` element that we can query using the [ByTitle](https://testing-library.com/docs/queries/bytitle) method. However, I don't have this offhand, but I do remember seeing an issue reported to React Testing Library regarding not being able to query `title` if it is not a direct child of the `svg` element, and this is exactly what we are having in our codebase.

https://github.com/freeCodeCamp/freeCodeCamp/blob/1e26dc1529cb9274c3a3edc2599d39c6cb84f714/client/src/assets/icons/green-pass.tsx#L12-L21

I'm okay with using `data-testid` in both the `svg` and the `div` 🙂

File: tools/challenge-helper-scripts/create-project.ts
Comment (huyenltnguyen) [inline tools/challenge-helper-scripts/create-project.ts:2]: Just wanted to be sure, is this change intentional?

**Review by huyenltnguyen — COMMENTED** [review]
Glad to see the PR passed CI now! Just some small changes and I think we are good to go.

File: tools/challenge-helper-scripts/create-project.ts
Comment (anirudh-modi) [reply tools/challenge-helper-scripts/create-project.ts:2]: Yes @huyenltnguyen it was indeed intentional.. there was an eslint error so to resolve it I had to change this code..

Comment (huyenltnguyen) [reply tools/challenge-helper-scripts/create-project.ts:2]: Could you help share the error and perhaps your Node version as well?

I find it odd that these ESLint plugins would touch non-test files.

Comment (anirudh-modi) [reply tools/challenge-helper-scripts/create-project.ts:2]: Oh this is weird, I remember getting an error now I am not getting.. I will revert the same..

File: client/src/components/settings/Certification.test.js
Comment (anirudh-modi) [reply client/src/components/settings/Certification.test.js]: Thanks @huyenltnguyen For the suggestion.. I have refractored and implemented your suggestions..

**Comment by anirudh-modi** [comment]
@huyenltnguyen I have refactored the code as per your suggestion.. Let me know if anything more needs to be changed or not.. 😄

**Review by huyenltnguyen — DISMISSED** [review]
@anirudh-modi Wow, thanks for addressing the feedback so quickly! The PR looks good to me. Thank you again for helping us with this effort 👍 

---
@freeCodeCamp/dev-team This PR enforces stricter rules for test files and when merged, other open PRs, especially those with unit tests, are required to update from the main branch and address ESLint errors, if any. I would advise hold off merging this one if you have PRs with higher priority or urgency. Feel free to summon me if I can be of any help 🙂

**Comment by anirudh-modi** [comment]
Thanks @huyenltnguyen for your guidance and suggestions.. I will be active on this PR, and if anyone has any other comments, I will resolve the same to ensure this PR can be merged..

File: client/src/components/Intro/Intro.test.js
Comment (ojeytonwilliams) [inline client/src/components/Intro/Intro.test.js:15]: Hi @anirudh-modi, this is a bit of weird one.  Unfortunately, the new lint rules are throwing false positives here.  They're interpreting `react-test-renderer` functions as `@testing-library/react` functions.

What I recommend is to revert these changes and instead leave a comment explaining why the lint rules had to be disabled for this case.

File: client/src/components/settings/Honesty.test.js
Comment (ojeytonwilliams) [inline client/src/components/settings/Honesty.test.js:28]: Same as above, could you revert the async/await changes and explain why these rules need to be disabled here?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hi @anirudh-modi, these tests look much improved.  If you could take a quick look at the false positives, then I think this is good to go.

File: client/src/components/Intro/Intro.test.js
Comment (huyenltnguyen) [reply client/src/components/Intro/Intro.test.js:15]: Thanks for catching this, Oliver. I missed this one.

**Comment by huyenltnguyen** [comment]
We should also update the branch to the main branch as well to be safe, since there have been a lot of changes going in recently.

File: client/src/components/Intro/Intro.test.js
Comment (ojeytonwilliams) [reply client/src/components/Intro/Intro.test.js:15]: No worries.  You've done an amazing job QAing this, so thanks again.

Comment (anirudh-modi) [reply client/src/components/Intro/Intro.test.js:15]: Hi @ojeytonwilliams thanks for your feedbacks, let me have a look at the same.. on the false postiive.. @huyenltnguyen Let me update my branch with `main` as well

File: client/src/components/settings/Honesty.test.js
Comment (anirudh-modi) [reply client/src/components/settings/Honesty.test.js]: Hi I have resolved the same

File: client/src/components/Intro/Intro.test.js
Comment (anirudh-modi) [reply client/src/components/Intro/Intro.test.js]: @huyenltnguyen I have also updated my branch with `main` and @ojeytonwilliams I have disabled the eslint for the given test with appropriate comment.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1: 

Great work @anirudh-modi :100:

**Review by huyenltnguyen — APPROVED** [review]

**Comment by huyenltnguyen** [comment]
Okay, I'm merging the PR now. Thanks again for the awesome work @anirudh-modi!

@freeCodeCamp/dev-team Please help note that open PRs are required to sync up with the main branch after this change is merged.

**Comment by anirudh-modi** [comment]
Thanks alot @huyenltnguyen.. really appreciate your guidance on this :) 
@ojeytonwilliams.. it was a pleasure to contribute..

**Pull request merged**
